File: webapp/backend/internal/models/famous_person.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/lib/pq"
	"gorm.io/gorm"
)

// FamousPerson category constants
const (
	FamousPersonCategoryPolitics   = "chinh_tri"
	FamousPersonCategoryScience    = "khoa_hoc"
	FamousPersonCategoryArt        = "nghe_thuat"
	FamousPersonCategoryMusic      = "am_nhac"
	FamousPersonCategorySport      = "the_thao"
	FamousPersonCategoryLiterature = "van_hoc"
	FamousPersonCategoryHistory    = "lich_su"
	FamousPersonCategoryFilm       = "dien_anh"
	FamousPersonCategoryBusiness   = "kinh_doanh"
	FamousPersonCategoryOther      = "khac"
)

// FamousPerson represents a notable historical or contemporary figure.
type FamousPerson struct {
	ID           uuid.UUID      `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	Name         string         `gorm:"type:varchar(255);not null" json:"name"`
	OriginalName string         `gorm:"type:varchar(255)" json:"original_name,omitempty"`
	BirthDate    *time.Time     `gorm:"type:date" json:"birth_date,omitempty"`
	BirthDay     *int           `gorm:"type:int" json:"birth_day,omitempty"`
	BirthMonth   *int           `gorm:"type:int" json:"birth_month,omitempty"`
	BirthYear    *int           `gorm:"type:int" json:"birth_year,omitempty"`
	DeathDate    *time.Time     `gorm:"type:date" json:"death_date,omitempty"`
	Nationality  string         `gorm:"type:varchar(100)" json:"nationality,omitempty"`
	Occupation   string         `gorm:"type:varchar(500)" json:"occupation,omitempty"`
	Category     string         `gorm:"type:varchar(50);not null;default:'khac'" json:"category"`
	ShortBio     string         `gorm:"type:text" json:"short_bio,omitempty"`
	ImageURL     string         `gorm:"type:varchar(500)" json:"image_url,omitempty"`
	ArticleID    *uuid.UUID     `gorm:"type:uuid;index" json:"article_id,omitempty"`
	IsVietnamese bool           `gorm:"default:false" json:"is_vietnamese"`
	Tags         pq.StringArray `gorm:"type:text[];default:'{}'" json:"tags,omitempty"`
	IsActive     bool           `gorm:"default:true" json:"is_active"`
	CreatedAt    time.Time      `gorm:"type:timestamptz;not null;default:now()" json:"created_at"`
	UpdatedAt    time.Time      `gorm:"type:timestamptz;not null;default:now()" json:"updated_at"`
	DeletedAt    gorm.DeletedAt `gorm:"type:timestamptz;index" json:"deleted_at,omitempty"`

	// Associations
	Article *Article `gorm:"foreignKey:ArticleID" json:"article,omitempty"`
}

func (FamousPerson) TableName() string {
	return "famous_people"
}

File: webapp/frontend/src/types/famousPerson.ts
// ============================================
// Famous Person Types
// ============================================

/**
 * Famous person data returned from the API
 */
export interface FamousPerson {
  id: string;
  name: string;
  original_name?: string;
  birth_date: string | null;
  birth_day?: number | null;
  birth_month?: number | null;
  birth_year?: number | null;
  death_date: string | null;
  nationality: string;
  occupation?: string;
  category: string;
  short_bio: string;
  image_url: string;
  tags: string[];
  is_vietnamese: boolean;
  is_active: boolean;
  article_id: string | null;
  created_at: string;
}

/**
 * Famous person summary (for list views)
 */
export interface FamousPersonSummary {
  id: string;
  name: string;
  original_name?: string;
  birth_date: string | null;
  death_date: string | null;
  nationality: string;
  short_bio: string;
  image_url: string;
  category: string;
  is_vietnamese: boolean;
  is_active: boolean;
  created_at: string;
}

/**
 * Create famous person request payload
 */
export interface CreateFamousPersonRequest {
  name: string;
  original_name?: string;
  birth_date?: string;
  death_date?: string;
  nationality?: string;
  occupation?: string;
  category: string;
  short_bio?: string;
  image_url?: string;
  tags?: string[];
  is_vietnamese?: boolean;
  article_id?: string;
}

/**
 * Update famous person request payload
 */
export interface UpdateFamousPersonRequest {
  name?: string;
  original_name?: string;
  birth_date?: string;
  death_date?: string;
  nationality?: string;
  occupation?: string;
  category?: string;
  short_bio?: string;
  image_url?: string;
  tags?: string[];
  is_vietnamese?: boolean;
  article_id?: string;
}

/**
 * Famous person list query parameters
 */
export interface FamousPersonListParams {
  page?: number;
  limit?: number;
  search?: string;
  sort_by?: string;
  sort_order?: "asc" | "desc";
  nationality?: string;
  is_vietnamese?: boolean;
  category?: string;
}
